crates/holochain/src/conductor/conductor.rs

```diff
@@ -3028,16 +3054,64 @@ mod misc_impls {
         /// Inject records into a source chain for a cell.
         /// If the records form a chain segment that can be "grafted" onto the existing chain, it will be.
         /// Otherwise, a new chain will be formed using the specified records.
+        //
+        // TODO: could integrate this better with [`SourceChain::flush`] when re-visit our workflows
+        //       and use them more consistently for moving data around
         pub async fn graft_records_onto_source_chain(
             self: Arc<Self>,
             cell_id: CellId,
             validate: bool,
             records: Vec<Record>,
         ) -> ConductorApiResult<()> {
-            graft_records_onto_source_chain::graft_records_onto_source_chain(
-                self, cell_id, validate, records,
+            let chain = self.get_agent_source_chain(&cell_id, true).await?;
+
+            let authored = chain.author_db().clone().into();
+            let dht = chain.dht_db().clone();
+            let cache = chain.dht_db_cache().clone();
+
+            let CellId(dna_hash, agent_pubkey) = &cell_id;
+            let ribosome = self.get_ribosome(dna_hash)?;
+
+            let space = self.get_or_create_space(dna_hash)?;
+            let op_hashes = chain
+                // XXX: probably unnecessary clone
+                .graft_records_onto_source_chain(records.clone())
+                .await?
+                .into_iter()
+                .map(|(hash, _basis)| hash)
+                .collect();
+            let workspace = space
+                .source_chain_workspace(self.keystore().clone(), agent_pubkey.clone())
+                .await?;
+            let network = self
+                .holochain_p2p()
+                .clone()
+                .into_dna(dna_hash.clone(), self.get_chc(&cell_id));
+
+            // Validate
+            if validate {
+                // Run the individual record validations.
+                crate::core::workflow::inline_validate_records(
+                    workspace.clone(),
+                    network.clone(),
+                    self.clone(),
+                    ribosome,
+                    records,
+                )
+                .await?;
+            }
+
+            // Integrate the ops.
+            // XXX: this is not checking for authorityship and just integrating everything we author,
+            //      which is wasteful of disk space in a sharded network, but not dangerous.
+            //      The reason for doing this is that when doing a CHC sync during genesis, we have not
+            //      yet joined the network, so we can't actually check authorityship at that time.
```

Comment:
I thought code elsewhere would check this as long as we're on a network? It's *only* genesis that doesn't do the check. It doesn't mean the check is entirely accurate until we build a picture of the network and adjust our arc accordingly but still worth doing here for conductors that run long enough for that to matter.

Reply:
Are you saying we should always check for authorityship? I actually did this without check because there is an error if running during genesis -- it's worse than just an inaccurate check. It seemed not worth restructuring things to make that error go away. What I could do is add a flag to this function to specify whether we should do a check or not, so that during normal chain flushes we can run with check, and during genesis, not. What about that?

Reply:
That sounds like a good solution to me. I'm fine that a few ops like genesis make their way into the DHT database optimistically. But for the majority of ops we can do the check, so that seems like we get most of the benefit if we do this

Reply:
Check added